contracts/cw1155-base/src/contract.rs

```diff
@@ -0,0 +1,1239 @@
+            |balance: Option<Uint128>| -> StdResult<_> { Ok(balance.unwrap_or_default() + amount) },
+        )?;
+    }
+
+    Ok(TransferEvent {
+        from,
+        to,
+        token_id,
+        amount,
+    })
+}
+
+/// returns true iff the sender can execute approve or reject on the contract
+fn check_can_approve(
+    deps: Deps,
+    env: &Env,
+    owner: &HumanAddr,
+    operator: &HumanAddr,
+) -> StdResult<bool> {
+    // owner can approve
+    let owner_raw = deps.api.canonical_address(owner)?;
+    let operator_raw = deps.api.canonical_address(operator)?;
+    if owner_raw == operator_raw {
+        return Ok(true);
+    }
+    // operator can approve
+    let op = APPROVES.may_load(deps.storage, (&owner_raw, &operator_raw))?;
+    Ok(match op {
+        Some(ex) => !ex.is_expired(&env.block),
+        None => false,
+    })
+}
+
+fn guard(cond: bool, err: ContractError) -> Result<(), ContractError> {
+    if cond {
+        Ok(())
+    } else {
+        Err(err)
+    }
+}
+
+pub fn execute_send_from(
+    env: ExecuteEnv,
+    from: HumanAddr,
+    to: HumanAddr,
+    token_id: TokenId,
+    amount: Uint128,
+    msg: Option<Binary>,
+) -> Result<Response, ContractError> {
+    let ExecuteEnv {
+        mut deps,
+        env,
+        info,
+    } = env;
+
+    guard(
+        check_can_approve(deps.as_ref(), &env, &from, &info.sender)?,
+        ContractError::Unauthorized {},
+    )?;
+
+    let mut rsp = Response::default();
+
+    let event = execute_transfer_inner(&mut deps, Some(&from), Some(&to), &token_id, amount)?;
+    event.add_attributes(&mut rsp);
+
+    rsp.messages = if let Some(msg) = msg {
+        vec![Cw1155ReceiveMsg {
+            operator: info.sender,
+            from: Some(from.clone()),
+            amount,
+            token_id: token_id.clone(),
+            msg,
+        }
+        .into_cosmos_msg(to)?]
+    } else {
+        vec![]
+    };
+
+    Ok(rsp)
+}
+
+pub fn execute_mint(
+    env: ExecuteEnv,
+    to: HumanAddr,
+    token_id: TokenId,
+    amount: Uint128,
+    msg: Option<Binary>,
+) -> Result<Response, ContractError> {
+    let ExecuteEnv { mut deps, info, .. } = env;
+
+    let sender = deps.api.canonical_address(&info.sender)?;
+    if sender != MINTER.load(deps.storage)? {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let mut rsp = Response::default();
+
+    let event = execute_transfer_inner(&mut deps, None, Some(&to), &token_id, amount)?;
+    event.add_attributes(&mut rsp);
+
+    rsp.messages = if let Some(msg) = msg {
+        vec![Cw1155ReceiveMsg {
+            operator: info.sender,
+            from: None,
+            amount,
+            token_id: token_id.clone(),
+            msg,
+        }
+        .into_cosmos_msg(to)?]
+    } else {
+        vec![]
+    };
+
+    // insert if not exist
+    let key = TOKENS.key(&token_id);
+    if deps.storage.get(&key).is_none() {
+        key.save(deps.storage, &"".to_owned())?;
```

Comment:
This means creating new token with no metadata?

Nice trick with `deps.storage.get(&key).is_none()` to do a "has" query and avoid parsing.
Should we support something like `TOKENS.has(&token_id)?` directly in cw-storage-plus?

Reply:
> This means creating new token with no metadata?

yes, insert an empty entry so token enumeration can find it.

> 
> Nice trick with `deps.storage.get(&key).is_none()` to do a "has" query and avoid parsing.
> Should we support something like `TOKENS.has(&token_id)?` directly in cw-storage-plus?

I think that'll be useful, another option is `key.exists(deps.storage)`.